clippy_lints/src/expr_metavars_in_unsafe.rs

```diff
@@ -0,0 +1,241 @@
+use std::collections::btree_map::Entry;
+use std::collections::BTreeMap;
+
+use clippy_utils::diagnostics::span_lint_hir_and_then;
+use clippy_utils::is_lint_allowed;
+use itertools::Itertools;
+use rustc_hir::intravisit::{walk_block, walk_expr, walk_stmt, Visitor};
+use rustc_hir::{BlockCheckMode, Expr, ExprKind, HirId, Stmt, UnsafeSource};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::impl_lint_pass;
+use rustc_span::def_id::DefId;
+use rustc_span::{sym, Span, SyntaxContext};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Looks for macros that expand metavariables in an unsafe block.
+    ///
+    /// ### Why is this bad?
+    /// This is unsound: it allows the user of the macro to write unsafe code outside of an
+    /// unsafe block at callsite, potentially invoking undefined behavior in safe code.
```

Comment:
> That in itself is not unsound

Hmm, I was going to refer back to [UCG](https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#soundness-of-code--of-a-library), which *to me* made this sound like it would clearly be unsound, practically speaking: the user doesn't have unsafe blocks in their code and the macro is in the public API of a library, but I guess it depends on whether you consider code inside of a macro-expanded unsafe block still be safe code.

According to https://github.com/rust-lang/unsafe-code-guidelines/issues/278 , it looks like it's not entirely clear if this is sound or not, so I guess it's better to just avoid mentioning soundness, then.

Reply:
I reworded this a bit in the last commit to avoid mentioning the terms (un)soundness. Lmk what you think